### DISCUSSION TITLE
3.1 branch divergence and future release plans

### DISCUSSION THREAD

**Discussion by davetapley** [falconry/falcon#2212] [General]
I just wanted to call out that the 3.1.x branches diverged way back on Fri Mar 25 2022 at 33c84505ab67664726d71d9fd34d0a8d55f02445, and so all the features since then aren't available 😑 

For me that's awkward because I really need:
- https://github.com/falconry/falcon/pull/1945

But it was merged June 1 2022 at edd9352e630dbbb6272370281fc5fa6d792df057, missing the cutoff 😞 

---

Is this intentional, if so is there a release plan?

Thanks for a great project 🙏🏻

**Comment by davetapley** [comment]
As an aside, I'm working around this right now with my fork, which is just 3.1.3 with edd9352e630dbbb6272370281fc5fa6d792df057 cherry picked in, and works great.

https://github.com/JEFuller/falcon/tree/3.1.3-multiple_segments

I thought about making a PR, but there isn't a sensible target for it (since `master` is diverged, and I _presume_ you want `3.1.3` to remain fixed).

**Comment by CaselIT** [comment]
Hi,

> I just wanted to call out that the 3.1.x branches diverged way back on Fri Mar 25 2022 at https://github.com/falconry/falcon/commit/33c84505ab67664726d71d9fd34d0a8d55f02445, and so all the features since then aren't available 😑

the main branch is the development version of the next version of falcon

> Is this intentional, if so is there a release plan?

Yes, that was a new feature that's planned for v4. Things are not progressing as fast as we would like toward a new release though.

**Comment by vytas7** [comment]
Yes @davetapley , we're just stretched on resources (as many open source projects... :imp:)...

The plan was to follow up 3.1 directly with 4.0. However, new CPython versions came out, requiring our immediate attention (IIRC the framework was completely broken under 3.11, 3.12 was not that critical, but we lacked binary wheels). So we just added patches atop the stable version (3.1).

The next version will finally be [4.0](https://github.com/falconry/falcon/milestone/34). We can maybe push some issues in favour of a timely release, but some we really need to address.

The roadmap for the 4.x series is outlined here: https://github.com/falconry/falcon/issues/2073.

**Comment by vytas7** [comment]
And thanks for your kind words about the project @davetapley!
Your fork with cherry-picked changes sounds like a great workaround for the time being. Alternatively, you could use just a snapshot of the `master` branch. Although it sounds adventurous as it is unstable/unreleased, but we normally do keep it in great shape too (with all CI passing, and full code coverage at all times).

**Comment by vytas7** [comment]
Thanks for your patience @davetapley, I just wanted to let you known that the wait was over, and the divergence is no more. [Falcon 4.0](https://github.com/falconry/falcon/releases/tag/4.0.0) was released last Friday. :sweat_smile:
